// x-pack/plugins/canvas/common/lib/fonts.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.americanTypewriter = {
    label: 'American Typewriter',
    value: "'American Typewriter', 'Courier New', Courier, Monaco, mono",
};
exports.arial = { label: 'Arial', value: 'Arial, sans-serif' };
exports.baskerville = {
    label: 'Baskerville',
    value: "Baskerville, Georgia, Garamond, 'Times New Roman', Times, serif",
};
exports.bookAntiqua = {
    label: 'Book Antiqua',
    value: "'Book Antiqua', Georgia, Garamond, 'Times New Roman', Times, serif",
};
exports.brushScript = {
    label: 'Brush Script',
    value: "'Brush Script MT', 'Comic Sans', sans-serif",
};
exports.chalkboard = { label: 'Chalkboard', value: "Chalkboard, 'Comic Sans', sans-serif" };
exports.didot = {
    label: 'Didot',
    value: "Didot, Georgia, Garamond, 'Times New Roman', Times, serif",
};
exports.futura = { label: 'Futura', value: 'Futura, Impact, Helvetica, Arial, sans-serif' };
exports.gillSans = {
    label: 'Gill Sans',
    value: "'Gill Sans', 'Lucida Grande', 'Lucida Sans Unicode', Verdana, Helvetica, Arial, sans-serif",
};
exports.helveticaNeue = {
    label: 'Helvetica Neue',
    value: "'Helvetica Neue', Helvetica, Arial, sans-serif",
};
exports.hoeflerText = {
    label: 'Hoefler Text',
    value: "'Hoefler Text', Garamond, Georgia, 'Times New Roman', Times, serif",
};
exports.lucidaGrande = {
    label: 'Lucida Grande',
    value: "'Lucida Grande', 'Lucida Sans Unicode', Lucida, Verdana, Helvetica, Arial, sans-serif",
};
exports.myriad = { label: 'Myriad', value: 'Myriad, Helvetica, Arial, sans-serif' };
exports.openSans = { label: 'Open Sans', value: "'Open Sans', Helvetica, Arial, sans-serif" };
exports.optima = {
    label: 'Optima',
    value: "Optima, 'Lucida Grande', 'Lucida Sans Unicode', Verdana, Helvetica, Arial, sans-serif",
};
exports.palatino = {
    label: 'Palatino',
    value: "Palatino, 'Book Antiqua', Georgia, Garamond, 'Times New Roman', Times, serif",
};
exports.fonts = [
    exports.americanTypewriter,
    exports.arial,
    exports.baskerville,
    exports.bookAntiqua,
    exports.brushScript,
    exports.chalkboard,
    exports.didot,
    exports.futura,
    exports.gillSans,
    exports.helveticaNeue,
    exports.hoeflerText,
    exports.lucidaGrande,
    exports.myriad,
    exports.openSans,
    exports.optima,
    exports.palatino,
];
